Pulumi 3.31.0 panics when trying to show some diff

### What happened?

Hi! :wave:

I’m hitting a weird panic in Pulumi 3.31.0 (it works in 3.30.0) when I’m trying to show a preview diff for a (new-to-my-stack) resource defined by the kubernetes.helm.v3.Chart for Sentry:

```
$ pulumi preview --diff

<the normal output beginnings, and a bunch of resource previews, until…>

panic: fatal: A failure has occurred: Unrecognized PropertyMap key type: reflect.Value

goroutine 55 [running]:
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.failfast(...)
	/private/tmp/pulumi-20220430-97257-15r72de/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v3/go/common/util/contract.Failf({0x5774bf1?, 0xc0006b7d00?}, {0xc0029b3950?, 0x203000?, 0x203000?})
	/private/tmp/pulumi-20220430-97257-15r72de/sdk/go/common/util/contract/fail.go:30 +0xcd
github.com/pulumi/pulumi/sdk/v3/go/common/resource.NewPropertyValueRepl({0x548efe0?, 0xc002842510}, 0x0, 0x0)
	/private/tmp/pulumi-20220430-97257-15r72de/sdk/go/common/resource/properties.go:344 +0xcda
github.com/pulumi/pulumi/sdk/v3/go/common/resource.NewPropertyValueRepl({0x54957c0?, 0xc002842480}, 0x0, 0x0)
	/private/tmp/pulumi-20220430-97257-15r72de/sdk/go/common/resource/properties.go:352 +0xd85
github.com/pulumi/pulumi/sdk/v3/go/common/resource.NewPropertyValueRepl({0x54957c0?, 0xc002842450}, 0x0, 0x0)
	/private/tmp/pulumi-20220430-97257-15r72de/sdk/go/common/resource/properties.go:352 +0xd85
github.com/pulumi/pulumi/sdk/v3/go/common/resource.NewPropertyValueRepl({0x54957c0?, 0xc0028423f0}, 0x0, 0x0)
	/private/tmp/pulumi-20220430-97257-15r72de/sdk/go/common/resource/properties.go:352 +0xd85
github.com/pulumi/pulumi/sdk/v3/go/common/resource.NewPropertyValue(...)
	/private/tmp/pulumi-20220430-97257-15r72de/sdk/go/common/resource/properties.go:261
github.com/pulumi/pulumi/pkg/v3/backend/display.(*propertyPrinter).decodeValue(0x53ec360?, {0xc0002c4e00?, 0xc0029b4190?})
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:1207 +0x9d
github.com/pulumi/pulumi/pkg/v3/backend/display.(*propertyPrinter).printPrimitivePropertyValue(0xc0029b4b80, {{0x53ec360?, 0xc0023aa990?}})
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:804 +0x1e5
github.com/pulumi/pulumi/pkg/v3/backend/display.(*propertyPrinter).printPropertyValue(0xc0029b4b80, {{0x53ec360?, 0xc0023aa990?}})
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:552 +0xcdc
github.com/pulumi/pulumi/pkg/v3/backend/display.(*propertyPrinter).printObjectProperty(0x53ec360?, {0xc00275c5b0?, 0xc00275c5b0?}, {{0x53ec360?, 0xc0023aa990?}}, 0xc0016381a0?)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:243 +0x4e
github.com/pulumi/pulumi/pkg/v3/backend/display.(*propertyPrinter).printObject(0xc0029b4b80, 0x1?)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:236 +0x10f
github.com/pulumi/pulumi/pkg/v3/backend/display.(*propertyPrinter).printPropertyValue(0xc0029b4dd0, {{0x5688320?, 0xc0025d5d10?}})
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:611 +0x9d2
github.com/pulumi/pulumi/pkg/v3/backend/display.(*propertyPrinter).printObjectProperty(0x5688320?, {0xc00275c56c?, 0xc00275c56c?}, {{0x5688320?, 0xc0025d5d10?}}, 0x30?)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:243 +0x4e
github.com/pulumi/pulumi/pkg/v3/backend/display.(*propertyPrinter).printObject(0xc0029b4dd0, 0x5af98e0?)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:236 +0x10f
github.com/pulumi/pulumi/pkg/v3/backend/display.PrintObject(...)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:225
github.com/pulumi/pulumi/pkg/v3/backend/display.getResourcePropertiesDetails({{0x572d275, 0x6}, {0xc0018c6bb0, 0xae}, {0xc001ef2000, 0x1c}, 0x0, 0xc00064c1b0, 0xc00064c240, {0x0, ...}, ...}, ...)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/object_diff.go:185 +0x1ce
github.com/pulumi/pulumi/pkg/v3/backend/display.renderDiff({_, _}, {{0x572d275, 0x6}, {0xc0018c6bb0, 0xae}, {0xc001ef2000, 0x1c}, 0x0, 0xc00064c1b0, ...}, ...)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/diff.go:326 +0x3f0
github.com/pulumi/pulumi/pkg/v3/backend/display.renderDiffResourcePreEvent({{{0x572d275, 0x6}, {0xc0018c6bb0, 0xae}, {0xc001ef2000, 0x1c}, 0x0, 0xc00064c1b0, 0xc00064c240, {0x0, ...}, ...}, ...}, ...)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/diff.go:347 +0x330
github.com/pulumi/pulumi/pkg/v3/backend/display.RenderDiffEvent({{0x573b7d6?, 0x0?}, {0x559d0c0?, 0xc001793a40?}}, 0x0?, {{0x572cef7, 0x6}, 0x0, 0x0, 0x0, ...})
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/diff.go:119 +0x4e5
github.com/pulumi/pulumi/pkg/v3/backend/display.ShowDiffEvents({0xc0012dc000, 0x11}, 0xc000950540, 0xc0009506c0, {{0x572cef7, 0x6}, 0x0, 0x0, 0x0, 0x0, ...})
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/diff.go:83 +0x5d0
github.com/pulumi/pulumi/pkg/v3/backend/display.ShowEvents({0xc0012dc000, 0x11}, {0x5730e01, 0x7}, {0x7ffeefbff6a2, 0xe}, {0xc0009697e0, 0xa}, 0xc000b81ef0?, 0xc0009506c0, ...)
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/display/display.go:59 +0x1fc
created by github.com/pulumi/pulumi/pkg/v3/backend/filestate.(*localBackend).apply
	/private/tmp/pulumi-20220430-97257-15r72de/pkg/backend/filestate/backend.go:557 +0x50f
```

### Steps to reproduce

I declare the Helm 3 Chart resource as follows:

```
    new kubernetes.helm.v3.Chart(
      'sentry',
      {
        namespace: 'REPLACEME',
        chart: 'sentry',
        fetchOpts: { repo: 'https://sentry-kubernetes.github.io/charts' },
…
```

This resource is the child of a ComponentResource, which is the child of another ComponentResource, which is… you get the idea. I have a bunch of nesting.

`$ pulumi preview` without `--diff` works fine, and `$ pulumi up` works until I select `details`, so I’m pretty certain it’s something to do with the diffing or preview logic. The stack trace also supports it.


### Expected Behavior

A diff preview of all the resources defined in the chart, like I get from version 3.30.0.

### Actual Behavior

Pulumi prints output and preview diff until it gets to the offending resource, then panics.

### Versions used

This one fails:

```
$ pulumi about

CLI
Version      3.31.0
Go Version   go1.18.1
Go Compiler  gc

Plugins
NAME    VERSION
nodejs  unknown

Host
OS       darwin
Version  11.6.2
Arch     x86_64

This project is written in nodejs (/Users/theneva/.nvm/versions/node/v16.14.0/bin/node v16.14.0)

Backend
Name           K.local
URL            gs://<redacted>
User           <redacted>
Organizations
```

This one works: 

```
$ pulumi about
CLI
Version      3.30.0
Go Version   go1.17.9
Go Compiler  gc

Plugins
NAME    VERSION
nodejs  unknown

Host
OS       darwin
Version  11.6.2
Arch     x86_64

This project is written in nodejs (/Users/theneva/.nvm/versions/node/v16.14.0/bin/node v16.14.0)

Backend
Name           K.local
URL            gs://<redacted>
User           <redacted>
Organizations
```

(Pulumi fails to get info about plugins because I’ve got my Pulumi project directories inside a yarn workspaces-based monorepo, which only has a `yarn.lock` a couple of directories up. It's another issue entirely, but it sure would be nice if Pulumi tried looking upwards to find a yarn.lock or package-lock.json file :smile:)

### Additional context

I suppose the issue could also be related to the go bump from 1.17 to 1.18 – I have no idea why the older version I installed with the install script uses 1.17, but the latest version installed with brew uses 1.18…

I’m running against a GCS backend, but I can’t imagine that it matters.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -4,3 +4,6 @@
 
 - [cli] Empty passphrases environment variables are now treated as if the variable was not set.
   [#9490](https://github.com/pulumi/pulumi/pull/9490)
+
+- [cli] Decode YAML mappings with numeric keys during diff.
+  [#9502](https://github.com/pulumi/pulumi/pull/9503)

--- a/pkg/backend/display/object_diff.go
+++ b/pkg/backend/display/object_diff.go
@@ -1194,7 +1194,11 @@ func (p *propertyPrinter) decodeValue(repr string) (resource.PropertyValue, stri
 
 		r.Reset(repr)
 		if err := yaml.NewDecoder(r).Decode(&object); err == nil {
-			return object, "yaml", true
+			translated, ok := p.translateYAMLValue(object)
+			if !ok {
+				return nil, "", false
+			}
+			return translated, "yaml", true
 		}
 
 		return nil, "", false
@@ -1208,4 +1212,59 @@ func (p *propertyPrinter) decodeValue(repr string) (resource.PropertyValue, stri
 		}
 	}
 	return resource.PropertyValue{}, "", false
+}
+
+// translateYAMLValue attempts to replace map[interface{}]interface{} values in a decoded YAML value with
+// map[string]interface{} values. map[interface{}]interface{} values can arise from YAML mappings with keys that are
+// not strings. This method only translates such maps if they have purely numeric keys--maps with slice or map keys
+// are not translated.
+func (p *propertyPrinter) translateYAMLValue(v interface{}) (interface{}, bool) {
+	switch v := v.(type) {
+	case []interface{}:
+		for i, e := range v {
+			ee, ok := p.translateYAMLValue(e)
+			if !ok {
+				return nil, false
+			}
+			v[i] = ee
+		}
+		return v, true
+	case map[string]interface{}:
+		for k, e := range v {
+			ee, ok := p.translateYAMLValue(e)
+			if !ok {
+				return nil, false
+			}
+			v[k] = ee
+		}
+		return v, true
+	case map[interface{}]interface{}:
+		vv := make(map[string]interface{}, len(v))
+		for k, e := range v {
+			sk := ""
+			switch k := k.(type) {
+			case string:
+				sk = k
+			case int:
+				sk = strconv.FormatInt(int64(k), 10)
+			case int64:
+				sk = strconv.FormatInt(k, 10)
+			case uint64:
+				sk = strconv.FormatUint(k, 10)
+			case float64:
+				sk = strconv.FormatFloat(k, 'g', -1, 64)
+			default:
+				return nil, false
+			}
+
+			ee, ok := p.translateYAMLValue(e)
+			if !ok {
+				return nil, false
+			}
+			vv[sk] = ee
+		}
+		return vv, true
+	default:
+		return v, true
+	}
 }

--- a/sdk/go/common/resource/properties.go
+++ b/sdk/go/common/resource/properties.go
@@ -333,23 +333,21 @@ func NewPropertyValueRepl(v interface{},
 	case reflect.Map:
 		// If a map, create a new property map, provided the keys and values are okay.
 		obj := PropertyMap{}
-		for _, key := range rv.MapKeys() {
-			var pk PropertyKey
-			switch k := key.Interface().(type) {
-			case string:
-				pk = PropertyKey(k)
-			case PropertyKey:
-				pk = k
-			default:
-				contract.Failf("Unrecognized PropertyMap key type: %v", reflect.TypeOf(key))
+		for iter := rv.MapRange(); iter.Next(); {
+			key := iter.Key()
+			if key.Kind() != reflect.String {
+				contract.Failf("Unrecognized PropertyMap key type %v", key.Type())
 			}
+
+			pk := PropertyKey(key.String())
 			if replk != nil {
 				if rk, repl := replk(string(pk)); repl {
 					pk = rk
 				}
 			}
-			val := rv.MapIndex(key)
-			pv := NewPropertyValueRepl(val.Interface(), replk, replv)
+
+			val := iter.Value().Interface()
+			pv := NewPropertyValueRepl(val, replk, replv)
 			obj[pk] = pv
 		}
 		return NewObjectProperty(obj)

